File: sample-conference-videochat/sample-conference-videochat/Components/Calls/ToolBar/ButtonsFactory.m
//
//  ButtonsFactory.m
//  sample-conference-videochat
//
//  Created by Injoit on 23/10/15.
//  Copyright © 2015 QuickBlox Team. All rights reserved.
//

#import "ButtonsFactory.h"

const CGRect kDefRect = {0, 0, 56, 76};
const CGRect kDefDeclineRect = {0, 0, 96, 44};
const CGRect kDefCircleDeclineRect = {0, 0, 44, 44};

#define kDefBackgroundColor [UIColor colorWithRed:0.85 green:0.89 blue:0.97 alpha:1.0]
#define kDefSelectedColor [UIColor colorWithRed:0.43 green:0.48 blue:0.58 alpha:1.0]
#define kDefDeclineColor [UIColor colorWithRed:0.94 green:0.20 blue:0.28 alpha:1.0]

@implementation ButtonsFactory

#pragma mark - Private

+ (CustomButton *)buttonWithFrame:(CGRect)frame
            backgroundColor:(UIColor *)backgroundColor
              selectedColor:(UIColor *)selectedColor
              selectedTitle:(NSString *)selectedTitle
            unSelectedTitle:(NSString *)unSelectedTitle {
    
    CustomButton *button = [[CustomButton alloc] initWithFrame:frame];
    button.selectedColor = selectedColor;
    button.unSelectedColor = backgroundColor;
    button.selectedTitle = selectedTitle;
    button.unSelectedTitle = unSelectedTitle;
    
    return button;
}

+ (UIImageView *)iconViewWithNormalImage:(NSString *)normalImage
                           selectedImage:(NSString *)selectedImage {
    
    UIImage *icon = [UIImage imageNamed:normalImage];
    UIImage *selectedIcon = [UIImage imageNamed:selectedImage];
    
    UIImageView *iconView = [[UIImageView alloc] initWithImage:icon
                                              highlightedImage:selectedIcon];
    
    iconView.contentMode = UIViewContentModeScaleAspectFit;
    
    return iconView;
}

#pragma mark - Public

+ (CustomButton *)videoEnable {
    CustomButton *button = [self buttonWithFrame:kDefRect
                           backgroundColor:kDefBackgroundColor
                             selectedColor:kDefSelectedColor selectedTitle:@"Cam on" unSelectedTitle:@"Cam off"];
    button.pushed = YES;
    button.iconView = [self iconViewWithNormalImage:@"camera_on_ic"
                                      selectedImage:@"cam_off"];
    return button;
}

+ (CustomButton *)auidoEnable {
    CustomButton *button = [self buttonWithFrame:kDefRect
                           backgroundColor:kDefBackgroundColor
                             selectedColor:kDefSelectedColor selectedTitle:@"Unmute" unSelectedTitle:@"Mute"];
    button.pushed = YES;
    button.iconView = [self iconViewWithNormalImage:@"mute_on_ic"
                                      selectedImage:@"mic_off"];
    return button;
}

+ (CustomButton *)screenShare {
    CustomButton *button = [self buttonWithFrame:kDefRect
                           backgroundColor:kDefBackgroundColor
                             selectedColor:kDefSelectedColor selectedTitle:@"Stop sharing" unSelectedTitle:@"Screen share"];
    button.iconView = [self iconViewWithNormalImage:@"screensharing_ic"
                                      selectedImage:@"screenshare_selected"];
    return button;
}

+ (CustomButton *)swapCam {
    CustomButton *button = [self buttonWithFrame:kDefRect
                           backgroundColor:kDefBackgroundColor
                             selectedColor:kDefSelectedColor selectedTitle:@"Swap cam" unSelectedTitle:@"Swap cam"];
    button.pushed = YES;
    button.iconView = [self iconViewWithNormalImage:@"switchCamera"
                                      selectedImage:@"abort_swap"];
    return button;
}

+ (CustomButton *)decline {
    CustomButton *button = [self buttonWithFrame:kDefRect
                           backgroundColor:kDefDeclineColor
                             selectedColor:kDefSelectedColor selectedTitle:@"End call" unSelectedTitle:@"End call"];
    button.pushed = YES;
    button.iconView = [self iconViewWithNormalImage:@"decline-ic"
                                      selectedImage:@"decline-ic"];
    return button;
}

@end

File: sample-conference-videochat-swift/sample-conference-videochat-swift/Components/Calls/ToolBar/ButtonsFactory.swift
//
//  ButtonsFactory.swift
//  sample-conference-videochat-swift
//
//  Created by Injoit on 04.10.2018.
//  Copyright © 2018 QuickBlox. All rights reserved.
//

import UIKit

struct ButtonsFactoryConstants {
    static let rect = CGRect(x: 0.0, y: 0.0, width: 56.0, height: 76.0)
    static let oldScreenRect = CGRect(x: 0.0, y: 0.0, width: 44.0, height: 44.0)
    
    static let backgroundColor = #colorLiteral(red: 0.8495520949, green: 0.8889414668, blue: 0.9678996205, alpha: 1)
    static let selectedColor = #colorLiteral(red: 0.4293757677, green: 0.4846315384, blue: 0.5802181363, alpha: 1)
    static let declineColor = #colorLiteral(red: 0.9356774688, green: 0.1954192817, blue: 0.2812524736, alpha: 1)
}

class ButtonsFactory {
    // MARK: - Class Methods
    class func button(withFrame frame: CGRect, backgroundColor: UIColor, selectedColor: UIColor, selectedTitle: String, unSelectedTitle: String) -> CustomButton {
        let button = CustomButton(frame: frame)
        button.selectedColor = selectedColor
        button.unSelectedColor = backgroundColor
        button.selectedTitle = selectedTitle
        button.unSelectedTitle = unSelectedTitle
        return button
    }
    
    class func iconView(withNormalImage normalImage: String, selectedImage: String) -> UIImageView? {
        let icon = UIImage(named: normalImage)
        let selectedIcon = UIImage(named: selectedImage)
        let iconView = UIImageView(image: icon, highlightedImage: selectedIcon)
        iconView.contentMode = .scaleAspectFit
        return iconView
    }
    
    class func videoEnable() -> CustomButton {
        let button: CustomButton = self.button(withFrame: ButtonsFactoryConstants.rect, backgroundColor:ButtonsFactoryConstants.backgroundColor, selectedColor: ButtonsFactoryConstants.selectedColor, selectedTitle: "Cam on", unSelectedTitle: "Cam off")
        button.pushed = true
        button.iconView = self.iconView(withNormalImage: "camera_on_ic", selectedImage: "cam_off")
        return button
    }
    
    class func audioEnable() -> CustomButton {
        let button: CustomButton = self.button(withFrame: ButtonsFactoryConstants.rect, backgroundColor: ButtonsFactoryConstants.backgroundColor, selectedColor: ButtonsFactoryConstants.selectedColor, selectedTitle: "Unmute", unSelectedTitle: "Mute")
        button.pushed = true
        button.iconView = self.iconView(withNormalImage: "mute_on_ic", selectedImage: "mic_off")
        return button
    }
    
    class func screenShare() -> CustomButton {
        let button: CustomButton = self.button(withFrame: ButtonsFactoryConstants.rect, backgroundColor: ButtonsFactoryConstants.backgroundColor, selectedColor: ButtonsFactoryConstants.selectedColor, selectedTitle: "Stop sharing", unSelectedTitle: "Screen share")
        button.iconView = self.iconView(withNormalImage: "screensharing_ic", selectedImage: "screenshare_selected")
        return button
    }
    
    class func swapCam() -> CustomButton {
        let button: CustomButton = self.button(withFrame: ButtonsFactoryConstants.rect, backgroundColor: ButtonsFactoryConstants.backgroundColor, selectedColor: ButtonsFactoryConstants.selectedColor, selectedTitle: "Swap cam", unSelectedTitle: "Swap cam")
        button.pushed = true
        button.iconView = self.iconView(withNormalImage: "switchCamera", selectedImage: "abort_swap")
        return button
    }
 
    class func decline() -> CustomButton {
        let button: CustomButton = self.button(withFrame: ButtonsFactoryConstants.rect, backgroundColor: ButtonsFactoryConstants.declineColor, selectedColor: ButtonsFactoryConstants.selectedColor, selectedTitle: "End call", unSelectedTitle: "End call")
        button.pushed = true
        button.iconView = self.iconView(withNormalImage: "decline-ic", selectedImage: "decline-ic")
        return button
    }
}
